{"TopicId":10492,"ForumId":29,"TopicTitle":"Simple GUI Question \u002B Try/Catch Question","CreatedByName":"OstermanA","CreatedById":8153,"CreatedDateTime":"2012-05-12T08:29:19Z","PageNum":1,"Articles":[{"CreatedByName":"OstermanA","CreatedById":8153,"CreatedDateTime":"2012-05-12T08:29:19Z","Content":"So, with Innsewerants\\\u0027 permission I have been tweaking the code for his mapping module. I have it behaving almost exactly as I want it to save for two annoying details.\n\nThe first is I have added a graphical toggle to turn the module on and off. I want it to default to center screen, and am not sure how to make it do this.\n\nThe other issue is I have added a bit of caching with the intention that if the output file is busy, it sill simply skip a write pass \u0026 wait for the next one. For some reason, the try/catch block I have implemented this with does not get the memo when something breaks. I\\\u0027ve tried using several commands to print debug messages with no dice, and the system log shows it dumping the entire cache even though it\\\u0027s supposed to abort. I am either misunderstanding how a try/catch works with a loop or the engine is not telling my script about the error.\n\nThe code can be found [here](http://ostermana.net/isa_RAM_Module.cs). If someone could take a look \u0026 give me a few pointers on where I screwed up, I\\\u0027d appreciate it.\n\nAlso, as I know there\\\u0027s at least one stupid thing in that, I should point out I don\\\u0027t actually know C#. I\\\u0027ve dabbled in Java and C\u002B\u002B, and am faking it."},{"CreatedByName":"Romfarer","CreatedById":20387,"CreatedDateTime":"2012-05-12T18:04:18Z","Content":"As for your first question to center the gui.\n\n    windowRect = GUI.Window (0, windowRect, WindowFunction, \u0027Map Module\u0027);\n\nYou can get the screen width and height by calling Screen.width and Screen.height. Divide it by 2 and you have the center position.\n\n    private Rect windowRect = new Rect(Screen.width/2, Screen.height/2, 100, 60);\n\nshoud suffice."},{"CreatedByName":"OstermanA","CreatedById":8153,"CreatedDateTime":"2012-05-14T06:17:44Z","Content":"So no one has any ideas why the try/catch is misbehaving?"},{"CreatedByName":"Romfarer","CreatedById":20387,"CreatedDateTime":"2012-05-14T07:30:05Z","Content":"I suspect your \u0027Using\u0027 statement might be used too narrowly. The StreamWriter object only lives inside the try block and thus it\\\u0027s exceptions are never caught."},{"CreatedByName":"FlutteryChicken","CreatedById":22399,"CreatedDateTime":"2012-05-14T16:30:49Z","Content":"I am assuming it is this try catch\n\n    try {\t\t\t\t\t\tusing (StreamWriter w = File.AppendText(kspDir \u002B \u0027isa_RAM_Mapper/isa_RAM_\u0027 \u002B bodyName \u002B \u0027_Data.csv\u0027)) {\t\t\t\t\t\t\twhile( isa_RAM_Buffer.Count \u003E 0 ) {\t\t\t\t\t\t\t\tw.WriteLine(isa_RAM_Buffer.Dequeue ());\t\t\t\t\t\t\t}\t\t\t\t\t\t\tw.Close ();\t\t\t\t\t\t\tisa_RAM_Pass_Count = 1;\t\t\t\t\t\t}\t\t\t\t\t} catch (IOException e) {\t\t\t\t\t\t\tConsole.WriteLine(\u0027File Write Error. Pass #\u0027 \u002B isa_RAM_Pass_Count \u002B \u0027 (\u0027 \u002B e \u002B \\\u0027)\\\u0027);\t\t\t\t\t\t\t\u002B\u002Bisa_RAM_Pass_Count;\t\t\t\t\t}\n\nif so, do this to check that it is actually an IOException by changing the IOException to Exception then using the ToString method to find out the actual cause of the exception.\n\nNote, do not leave it as exception, once you know what is can cause it, use the correct exception.\n\n    try {\t\t\t\t\t\tusing (StreamWriter w = File.AppendText(kspDir \u002B \u0027isa_RAM_Mapper/isa_RAM_\u0027 \u002B bodyName \u002B \u0027_Data.csv\u0027)) {\t\t\t\t\t\t\twhile( isa_RAM_Buffer.Count \u003E 0 ) {\t\t\t\t\t\t\t\tw.WriteLine(isa_RAM_Buffer.Dequeue ());\t\t\t\t\t\t\t}\t\t\t\t\t\t\tw.Close ();\t\t\t\t\t\t\tisa_RAM_Pass_Count = 1;\t\t\t\t\t\t}\t\t\t\t\t} catch (Exception e) {Console.WriteLine(e.ToString());\t\t\t\t\t\t\t//Console.WriteLine(\u0027File Write Error. Pass #\u0027 \u002B isa_RAM_Pass_Count \u002B \u0027 (\u0027 \u002B e \u002B \\\u0027)\\\u0027);\t\t\t\t\t\t\t//\u002B\u002Bisa_RAM_Pass_Count;\t\t\t\t\t}"},{"CreatedByName":"OstermanA","CreatedById":8153,"CreatedDateTime":"2012-05-17T09:31:48Z","Content":"I found the problem. The exception being caught was wrong. Should have been \u0027} catch (UnauthorizedAccessException uae) {\u0027. For S\u0026G\\\u0027s I added an additional \u0027} catch (PathTooLongException ptle) {\u0027 that shuts it down \u0026 clears the buffer."},{"CreatedByName":"N3X15","CreatedById":12073,"CreatedDateTime":"2012-05-27T16:40:42Z","Content":"You can\\\u0027t catch IOException yet, that\\\u0027s in the next release cycle."}]}